### PR TITLE
make `import.vis` is immutable

### PR DESCRIPTION
**Pull request by bvanjoi**
r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_resolve/src/imports.rs [hunk 1]
```diff
@@ -175,7 +175,7 @@ pub(crate) struct ImportData<'a> {
     pub module_path: Vec<Segment>,
     /// The resolution of `module_path`.
     pub imported_module: Cell<Option<ModuleOrUniformRoot<'a>>>,
-    pub vis: Cell<Option<ty::Visibility>>,
+    pub vis: Option<ty::Visibility>,
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/imports.rs:178]:
```suggestion
    pub vis: ty::Visibility,
```
This visibility should never be `None` now.
(That means `expect_vis` can also be removed.)

File: compiler/rustc_resolve/src/ident.rs [hunk 2]
```diff
@@ -1422,17 +1458,27 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         path: &[Segment],
         opt_ns: Option<Namespace>, // `None` indicates a module path in import
         parent_scope: &ParentScope<'a>,
+        used_import: UsedImport<'a>,
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/ident.rs:1461]:
```suggestion
        ignore_import: Option<Import<'a>>,
```
- This argument is very similar to `ignore_binding`, except here we ignore the import, so it is reasonable to name it similarly.
- Also, could move all these new arguments after the `ignore_binding` arguments (or otherwise to the last position).
- Also, the type alias doesn't seem very useful.

File: compiler/rustc_resolve/src/ident.rs [hunk 3]
```diff
@@ -1546,6 +1593,7 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                     ident,
                     ns,
                     parent_scope,
+                    used_import,
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/ident.rs:1596]: The import should always be `None` here, and `fn resolve_ident_in_lexical_scope` doesn't need the import argument.
`ribs.is_some()` means `used_import` is `None`, you can add an assert.

File: compiler/rustc_resolve/src/macros.rs [hunk 4]
```diff
@@ -788,6 +802,7 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
             self.report_out_of_scope_macro_calls(
                 ast_path,
                 parent_scope,
+                used_import,
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/macros.rs:805]: This is unnecessary, `invoc_in_mod_inert_attr.is_some()` means `used_import.is_none()`, you can add an assert.

File: compiler/rustc_resolve/src/build_reduced_graph.rs [hunk 5]
```diff
@@ -1369,7 +1371,6 @@ impl<'a, 'b, 'tcx> Visitor<'b> for BuildReducedGraphVisitor<'a, 'b, 'tcx> {
         self.parent_scope.module = orig_current_module;
         self.parent_scope.macro_rules = orig_current_macro_rules_scope;
     }
-
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/build_reduced_graph.rs:1372]: Unnecessary change.

File: compiler/rustc_resolve/src/ident.rs [hunk 6]
```diff
@@ -962,26 +987,29 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
...
         // if it can then our result is not determined and can be invalidated.
         for single_import in &resolution.single_imports {
-            let Some(import_vis) = single_import.vis.get() else {
-                // This branch handles a cycle in single imports, which occurs
-                // when we've previously **steal** the `vis` value during an import
-                // process.
+            let Some(import_vis) = single_import.vis else {
+                continue;
+            };
+            if !self.is_accessible_from(import_vis, parent_scope.module) {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/ident.rs:993]: It's better to move this check to its old place, the ignored import check is cheaper.

File: compiler/rustc_resolve/src/ident.rs [hunk 7]
```diff
@@ -962,26 +987,29 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
...
+                continue;
+            }
+            if let Some(used_import) = used_import
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/ident.rs:996]:
```suggestion
            if ignored_import == Some(single_import) {
```
And same below.

**Comment by petrochenkov** [comment]
So I wondered if it's possible to have multiple ignored imports at the same time, but it looks like it's not possible.
`vis.take()` was only called from `resolve_import` and `finalize_import` which are "top level" functions never called while resolving other imports.

File: compiler/rustc_resolve/src/ident.rs (hunk 3 above)
Comment (bvanjoi) [reply compiler/rustc_resolve/src/ident.rs:1596]: I will remove `used_import` in `resolve_ident_in_lexical_scope`.

File: compiler/rustc_resolve/src/macros.rs (hunk 4 above)
Comment (bvanjoi) [reply compiler/rustc_resolve/src/macros.rs:805]: I will remove `used_import` in `report_out_of_scope_macro_calls`.

File: compiler/rustc_resolve/src/imports.rs [hunk 8]
```diff
@@ -1013,7 +1007,7 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                     }
                     if !is_prelude
                         && let Some(max_vis) = max_vis.get()
-                        && let import_vis = import.expect_vis()
+                        && let import_vis = import.vis
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/imports.rs:1010]: Separate `import_vis` can be removed now.

**Comment by petrochenkov** [comment]
r=me with https://github.com/rust-lang/rust/pull/128732#discussion_r1706975850 addressed.
@bors delegate+

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 9]
```diff
@@ -2510,7 +2518,7 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
 
     /// Finds a cfg-ed out item inside `module` with the matching name.
     pub(crate) fn find_cfg_stripped(
-        &mut self,
+        &self,
```
Comment (bvanjoi) [inline compiler/rustc_resolve/src/diagnostics.rs:2521]: Just a note: there has been a small change from `&mut self` to `&self`.

**Comment by bvanjoi** [comment]
ci is green

@bors r=@petrochenkov
